lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,768 @@
+
+  defp validate_dynamic(:infinity), do: :ok
+  defp validate_dynamic(dynamic) when is_integer(dynamic) and dynamic >= 0, do: :ok
+  defp validate_dynamic(dynamic), do: {:error, {:invalid_max_dynamic, dynamic}}
+
+  @impl true
+  def handle_call(:which_children, _from, state) do
+    %{children: children} = state
+
+    reply =
+      for {pid, args} <- children do
+        case args do
+          {:restarting, {_, _, _, type, modules}} ->
+            {:undefined, :restarting, type, modules}
+
+          {_, _, _, type, modules} ->
+            {:undefined, pid, type, modules}
+        end
+      end
+
+    {:reply, reply, state}
+  end
+
+  def handle_call(:count_children, _from, state) do
+    %{children: children} = state
+    specs = map_size(children)
+
+    {active, workers, supervisors} =
+      Enum.reduce(children, {0, 0, 0}, fn
+        {_pid, {:restarting, {_, _, _, :worker, _}}}, {active, worker, supervisor} ->
+          {active, worker + 1, supervisor}
+
+        {_pid, {:restarting, {_, _, _, :supervisor, _}}}, {active, worker, supervisor} ->
+          {active, worker, supervisor + 1}
+
+        {_pid, {_, _, _, :worker, _}}, {active, worker, supervisor} ->
+          {active + 1, worker + 1, supervisor}
+
+        {_pid, {_, _, _, :supervisor, _}}, {active, worker, supervisor} ->
+          {active + 1, worker, supervisor + 1}
+      end)
+
+    reply = %{specs: specs, active: active, workers: workers, supervisors: supervisors}
+    {:reply, reply, state}
+  end
+
+  def handle_call({:terminate_child, pid}, _from, %{children: children} = state) do
+    case children do
+      %{^pid => info} ->
+        :ok = terminate_children(%{pid => info}, state)
+        {:reply, :ok, delete_child(pid, state)}
+
+      %{} ->
+        {:reply, {:error, :not_found}, state}
+    end
+  end
+
+  def handle_call({:start_child, child}, _from, state) do
+    %{dynamic: dynamic, max_dynamic: max_dynamic} = state
+
+    if dynamic < max_dynamic do
+      handle_start_child(child, %{state | dynamic: dynamic + 1})
+    else
+      {:reply, {:error, :max_dynamic}, state}
+    end
+  end
+
+  defp handle_start_child({{m, f, args} = mfa, restart, shutdown, type, modules}, state) do
+    case reply = start_child(m, f, args) do
+      {:ok, pid, _} ->
+        {:reply, reply, save_child(pid, mfa, restart, shutdown, type, modules, state)}
+
+      {:ok, pid} ->
+        {:reply, reply, save_child(pid, mfa, restart, shutdown, type, modules, state)}
+
+      _ ->
+        {:reply, reply, update_in(state.dynamic, &(&1 - 1))}
+    end
+  end
+
+  defp start_child(m, f, a) do
+    try do
+      apply(m, f, a)
+    catch
+      kind, reason ->
+        {:error, exit_reason(kind, reason, System.stacktrace())}
+    else
+      {:ok, pid, extra} when is_pid(pid) -> {:ok, pid, extra}
+      {:ok, pid} when is_pid(pid) -> {:ok, pid}
+      :ignore -> :ignore
+      {:error, _} = error -> error
+      other -> {:error, other}
```

Comment:
Do you mean `{:error, {:bad_return_value, other}}`?

Reply:
I am fine with wrapping in bad_return_value but as you mentioned above, we are being consistent with `Supervisor` here and this is not quite a `bad_return_value` since error is a valid return.

Reply:
Error isn't the return value and I meant `other -> {:error, {:bad_return_value, other}}`. The second paragraph was mentioning other inconsistencies. I am unsure if any of these would unexpected break anyones code...